File: my-folder/0907-koko-eating-bananas/solution.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        min_eat = float("inf")
        while (l <= r):
            mid_k = (l + r) // 2
            hours = 0
            for p in piles:
                hours += math.ceil(p/mid_k)
            if (hours > h):
                l = mid_k + 1
            else:
                min_eat = min(mid_k, min_eat)
                r = mid_k - 1
        
        return min_eat

# [3,6,7,11]
#  l = 1 

File: my-folder/0907-koko-eating-bananas/solution.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
    int currMax = 0;
    for (int pile : piles) {
      currMax = Math.max(currMax, pile);
    }

    int left = 1;
    int right = currMax;

    while (left < right) {
      int mid = left + (right - left) / 2;
      int currTimeTaken = 0;

      for (int pile : piles) {
        currTimeTaken += (pile + mid - 1) / mid; // ceiling division
      }

      if (currTimeTaken <= h) {
        right = mid; // try smaller speed
      } else {
        left = mid + 1; // need faster
      }
    }

    return left; 

    }    

}
